AB-3845 Add warning state for swedbank

simonbergstrom opened the pull request:
Since some special cases with Swedbank accounts with specific clearing
numbers occurs that will break checksum check and give false negatives
these errors are suppressed as warnings instead of errors. The function
will return these account numbers as valid with a warning flag and a
warning array on the bank object.

AB-3845

simonbergstrom committed:
Enable warning state for Swedbank accounts

Since some special cases with Swedbank accounts with specific clearing
numbers occurs that will break checksum check and give false negatives
these errors are suppressed as warnings instead of errors. The function
will return these account numbers as valid with a warning flag and a
warning array on the bank object.

AB-3845

alexteg-advisa commented on kontonummer.js:
```diff
@@ -432,20 +438,26 @@
             bank_name = bank.name;
             if (!(bank.modulo === 11 && mod11(ctrlNum)) && !(bank.modulo === 10 && mod10(ctrlNum))) {
-                errors.push('bad_checksum');
+                if(bank.warnOnBadChecksum) {
```
We don't have linting on this repo yet (since it was forked), but let's try to keep consistent:

```suggestion
                if (bank.warnOnBadChecksum) {
```

alexteg-advisa commented on kontonummer.js:
```diff
@@ -348,7 +348,8 @@
         },
-        zerofill: true
+        zerofill: true,
+        warnOnBadChecksum: true, // Special swedbank accounts can cause checksum errors which we conditionally want to suppress
```
While we may not have to care about [<IE9 support](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas#Browser_compatibility), this is a public repo that doesn't have any build step and seemingly trailing commas have been avoided, likely for compatibility reasons. I'd suggest to avoid trailing commas in this repo. This applies to a few other places in this PR.

```suggestion
        warnOnBadChecksum: true // Special swedbank accounts can cause checksum errors which we conditionally want to suppress
```

alexteg-advisa commented on kontonummer.js:
```diff
@@ -373,19 +374,22 @@
             errors: [ 'invalid_account_number' ],
-            matched_banks: []
+            matched_banks: [],
+            has_warnings: false,
```
Trailing comma

```suggestion
            has_warnings: false
```

alexteg-advisa commented on kontonummer.js:
```diff
@@ -404,14 +408,16 @@
                 errors: errors,
-                matched_banks: matched_banks
+                matched_banks: matched_banks,
+                has_warnings: Boolean(warnings.length),
```
Trailing comma

```suggestion
                has_warnings: Boolean(warnings.length)
```

alexteg-advisa commented on kontonummer.js:
```diff
@@ -404,14 +408,16 @@
             errors: [],
-            matched_banks: matched_banks
+            matched_banks: matched_banks,
+            has_warnings: Boolean(warnings.length),
```
Trailing comma

```suggestion
            has_warnings: Boolean(warnings.length)
```

alexteg-advisa commented on tests/kontonummer.js:
```diff
@@ -116,11 +117,39 @@ describe('Kontonummer suite', () => {
+        },
+        zerofill: true,
+        warnOnBadChecksum: true,
+    };
```
Incorrect indentation, should be one step more:

```suggestion
      };
```

alexteg-advisa commented on tests/kontonummer.js:
```diff
@@ -116,11 +117,39 @@ describe('Kontonummer suite', () => {
+        lengths : {
+            clearing : 5,
+            account  : 10,
+            control  : 10
```
Indentation of these 3 lines is not matching the rest of this file (2 spaces).

alexteg-advisa reviewed (changes requested):
My main concern is that we still don't really validate Swedbank account numbers _length_. Yes, we [validate the length](https://github.com/Advisa/kontonummer.js/blob/c648048a02aef0e6c7efbafb2cfff49da38edfb7/kontonummer.js#L391), but we validate (correctly) the [zeropadded account number](https://github.com/Advisa/kontonummer.js/blob/c648048a02aef0e6c7efbafb2cfff49da38edfb7/kontonummer.js#L388) and the Swedbank account number has [`zerofill: true`](https://github.com/Advisa/kontonummer.js/blob/c648048a02aef0e6c7efbafb2cfff49da38edfb7/kontonummer.js#L351). This means that an account number like `'80000'` are considered valid with a warning and an account number like `8000018` is valid both without errors and without warnings, which should be incorrect.

The core question here is what the shortest valid Swedbank account number with a `8*` clearing number is. This does not seem to be readily documented.
- [Bankgiro documentation](https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf) has no mention of zero padding at all
- [Swedbank](https://www.swedbank.se/privat/kort-och-betala/konton-for-in-och-utbetalningar/clearingnummer.html) doesn't mention it on their site.
- [Nordea](https://www.nordea.se/Images/39-112644/F%C3%B6rteckning%20clearing-%20och%20bankkontonummer.pdf) mentions that
  > "Clearingnummer numeriskt vänsterställt. Kontonummer numeriskt högerställt. Positionerna emellan nollutfyllda.", but nothing regarding shortest account number.
- [SEB](https://seb.se/privat/verktyg-for-din-ekonomi/sa-har-anvander-du-bank-pa-telefon) says 
  > "Swedbank: Om clearingnumret börjar på 7 skriver du 11 siffror. Om det börjar på 8 skriver du 15. Om antalet siffror är färre fyller du på med nollor mellan clearingnumret (de 5 första siffrorna) och kontonumret."
- [Danske Bank](https://danskebank.se/sv-se/privat/kontopaket/betala/betala-rakningar/betala-rakningar-i-hembanken/pages/bankernas-kontonummer-clearingnummer.aspx) says:
  > "När du gör en överföring till Swedbank ska kontonummer som börjar på 8 alltid bestå av sammanlagt fjorton siffror. Om det innehåller färre siffror ska du fylla ut med nollor mellan clearingnumret (de första fyra siffrorna) och resterande siffror i kontonumret."
- [DBSchenker](https://www.dbschenker.com/se-sv/haer-ser-du-hur-kontonumret-ska-skrivas-foer-olika-banker-481736) says:
  > "Om antalet siffror är färre än 15 fyller du på med nollor mellan clearingnumret och kontonumret."
- [Ikano](https://secure.ikanobank.se/site_files/docs/svenskabankers_kontonr.pdf) says:
  > "När du anger ett konto som börjar på 8 ska kontonumret alltid bestå av sammanlagt femton siffror. Om det innehåller färre siffror ska du fylla ut med nollor mellan clearingnumret (de fem första siffrorna) och resterande siffror i kontonumret."

So I called Swedbank customer service and they said they don't have any documentation regarding the shortest possible account number (somebody might of course have it, but customer service doesn't). However, they said the shortest they have seen is 7 digits in the account number part. I asked her to verify some account numbers we have with 5 digits, but their internal system declined to accept them all giving a validation error of incorrect clearing number (even though that was correct), while some account numbers with 6 characters were not found, but did not result in a validation error.

This reverse engineering suggests that the shortest possible account number Swedbank accepts is 6 digits (+ 5 digits of clearing number). I'd suggest to add validation for that in accordance to the original issue description to avoid accepting very short and invalid account numbers.

Otherwise just a few comments that having linting in this repo should have caught.

simonbergstrom commented on kontonummer.js:
```diff
@@ -348,7 +348,8 @@
             account  : 10,
```
Remove zero padding and set account to 6

simonbergstrom commented:
Great work with the research about the length of account numbers. I understand the issue but just want to clarify the possible update for this. So should I remove the zeropadding for the Swedbank account check and change the account according to my suggested changes?

alexteg-advisa commented:
@simonbergstrom Zeropadding is correct as mentioned ("we validate (correctly) the zeropadded account number"), so that shouldn't be changed, nor that the checksum is made on the zeropadded number (which it has to be, since it assumes a certain number of digits from the end of the combined clearing & zeropadded account number).

I'd argue the length check of the zeropadded number is also technically correct for all the other banks, even though it in this Swedbank case will always pass, since the zeros are added. Basically the `lengths.account` and `lengths.control` numbers refer to how long the bank account number should be when applying the zero-padding, but we lack a validation for the minimum number of significant, non-zero-prefixed, digits in the account number.

Considering the customer _might_ already provide none, some or all of the zero-padding, I'd suggest the following procedure:

1. Add a key like `min_account` to the `lengths` object of the [Swedbank `8*` clearing number group](https://github.com/Advisa/kontonummer.js/blob/c648048a02aef0e6c7efbafb2cfff49da38edfb7/kontonummer.js#L347-L349)
2. Use the [zero padded account number](https://github.com/Advisa/kontonummer.js/blob/c648048a02aef0e6c7efbafb2cfff49da38edfb7/kontonummer.js#L388) in the `validateLength` function and parse it as an integer to get rid of the zeros, turn it back into a string and in [this if](https://github.com/Advisa/kontonummer.js/blob/c648048a02aef0e6c7efbafb2cfff49da38edfb7/kontonummer.js#L428) check if the length is shorter than the `lengths.minimumAccount` so it results in the error `errors.push('too_short');` getting added.

simonbergstrom committed:
Add support for min account nr length validation

Add support for validate length for special case when checksum returns
warning a strict length check needs to be done. Since zero padding will
always give a valid length we need to perform the validation with no
padding.

AB-3845

alexteg-advisa commented on kontonummer.js:
```diff
@@ -404,24 +409,38 @@
+            
+            if(bank.lengths.min_account) {
+                // Since account number is zero padded we need to slice it before trim
+                const onlyAccountNumber = actualBankNumber.slice(bank.lengths.clearing);
```
There's actually a function to just get the account number `getAccountNumber`:

```suggestion
                const onlyAccountNumber = getAccountNumber(actualBankNumber, bank.lengths.clearing);
```

alexteg-advisa commented on kontonummer.js:
```diff
@@ -404,24 +409,38 @@
+            let actualBankNumber = bankNumber;
+            let minBankNumberLength =  bank.lengths.clearing + bank.lengths.account;
+            const maxBankNumberLength = bank.lengths.clearing + bank.lengths.account;
+            
```
Linting - avoid trailing whitespace:

```suggestion

```

alexteg-advisa commented on kontonummer.js:
```diff
@@ -404,24 +409,38 @@
+                const onlyAccountNumber = actualBankNumber.slice(bank.lengths.clearing);
+                const onlyClearingNumber = actualBankNumber.substr(0, bank.lengths.clearing);
+                actualBankNumber = `${onlyClearingNumber}${parseInt(onlyAccountNumber, 10)}`;
+                minBankNumberLength = bank.lengths.clearing + bank.lengths.min_account; 
```
`bank.lengths.min_account` only refers to the account number, not the clearing number.

Linting - avoid trailing whitespace:

```suggestion
                minBankNumberLength = bank.lengths.min_account;
```

alexteg-advisa commented on kontonummer.js:
```diff
@@ -404,24 +409,38 @@
+            let minBankNumberLength =  bank.lengths.clearing + bank.lengths.account;
+            const maxBankNumberLength = bank.lengths.clearing + bank.lengths.account;
+            
+            if(bank.lengths.min_account) {
```
Linting - space between `if` and parenthesis:

```suggestion
            if (bank.lengths.min_account) {
```

alexteg-advisa commented on kontonummer.js:
```diff
@@ -404,24 +409,38 @@
-            if (bankNumber.length < bank.lengths.clearing + bank.lengths.account) {
+            let actualBankNumber = bankNumber;
+            let minBankNumberLength =  bank.lengths.clearing + bank.lengths.account;
+            const maxBankNumberLength = bank.lengths.clearing + bank.lengths.account;
```
This doesn't have a babel/ES6 compilation/build step, so we should aim to have it ES5 compatible, so no `const` or `let`.

```suggestion
            var maxBankNumberLength = bank.lengths.clearing + bank.lengths.account;
```

alexteg-advisa commented on kontonummer.js:
```diff
@@ -404,24 +409,38 @@
+            if(bank.lengths.min_account) {
+                // Since account number is zero padded we need to slice it before trim
+                const onlyAccountNumber = actualBankNumber.slice(bank.lengths.clearing);
+                const onlyClearingNumber = actualBankNumber.substr(0, bank.lengths.clearing);
```
No need to get the clearing number, since we only need to check the account number length without the clearing number (but there is a function to get the clearing number):
```suggestion
```

alexteg-advisa commented on kontonummer.js:
```diff
@@ -404,24 +409,38 @@
+                // Since account number is zero padded we need to slice it before trim
+                const onlyAccountNumber = actualBankNumber.slice(bank.lengths.clearing);
+                const onlyClearingNumber = actualBankNumber.substr(0, bank.lengths.clearing);
+                actualBankNumber = `${onlyClearingNumber}${parseInt(onlyAccountNumber, 10)}`;
```
Clearing number should not be included when comparing with `bank.lengths.min_account`.

```suggestion
                actualBankNumber = `${parseInt(onlyAccountNumber, 10)}`;
```

alexteg-advisa commented on kontonummer.js:
```diff
@@ -404,24 +409,38 @@
+            const maxBankNumberLength = bank.lengths.clearing + bank.lengths.account;
+            
+            if(bank.lengths.min_account) {
+                // Since account number is zero padded we need to slice it before trim
```
Isn't it that we need to slice it because it contains both clearing number and account number, rather than because it is zero padded?

If we can avoid comments by writing more code with more expressive behaviour and naming, that is to be preferred. In this case we could create a function that removes the zero padding and name it something like `removeNumberZeroPadding`.

alexteg-advisa reviewed (approved):
I've pushed up a fixup commit that fixes these comments. With these fixes I think this is good to be squashed & merged, but @simonbergstrom review my fixes and tell me if there is something more you think should be fixed.
